velox/vector/tests/VectorTest.cpp

```diff
@@ -1671,3 +1671,64 @@ TEST_F(VectorTest, multipleDictionariesOverLazy) {
     ASSERT_EQ(i, dict->as<SimpleVector<int32_t>>()->valueAt(i));
   }
 }
+
+class LazyTestingLoader : public VectorLoader {
+ public:
+  explicit LazyTestingLoader(RowVectorPtr data, int32_t* rowCounter)
+      : data_(data) {
+    rowCounter_ = rowCounter;
+  }
+
+  void loadInternal(RowSet rows, ValueHook* hook, VectorPtr* result) override {
+    (*rowCounter_) += rows.size();
+    *result = data_;
+  }
+
+ private:
+  RowVectorPtr data_;
+  int32_t* rowCounter_;
+};
+
+/// Test lazy loading of dictionary vector
+TEST_F(VectorTest, selectiveLoadingOfLazyDictionary) {
+  auto vectorMaker = std::make_unique<test::VectorMaker>(pool_.get());
+
+  vector_size_t size = 10;
+  auto indices =
+      makeIndices(size, [&](auto row) { return (row % 2 == 0) ? row : size; });
+  int32_t rowCounter = 0;
+  auto rowData = makeRowVector(
+      vectorMaker->flatVector<int32_t>(size, [](auto row) { return row; }));
+  auto lazyVector = std::make_shared<LazyVector>(
+      pool_.get(),
+      INTEGER(),
+      size,
+      std::make_unique<LazyTestingLoader>(rowData, &rowCounter));
+
+  auto dict = BaseVector::wrapInDictionary(nullptr, indices, size, lazyVector);
+  dict->loadedVector();
+
+  ASSERT_EQ(rowCounter, size / 2);
+}
+
+/// Test lazy loading of nested dictionary vector
+TEST_F(VectorTest, selectiveLoadingOfLazyDictionaryNested) {
+  auto vectorMaker = std::make_unique<test::VectorMaker>(pool_.get());
+
+  vector_size_t size = 10;
+  auto indices =
+      makeIndices(size, [&](auto row) { return (row % 2 == 0) ? row : size; });
+  int32_t rowCounter = 0;
+  auto rowData = makeRowVector(
+      vectorMaker->flatVector<int32_t>(size, [](auto row) { return row; }));
+  auto lazyVector = std::make_shared<LazyVector>(
+      pool_.get(),
+      INTEGER(),
+      size,
+      std::make_unique<LazyTestingLoader>(rowData, &rowCounter));
+
+  auto dict = BaseVector::wrapInDictionary(nullptr, indices, size, BaseVector::wrapInDictionary(nullptr, indices, size, lazyVector));
+  dict->loadedVector();
```

Comment:
Can we add a test where calling loadedVector() twice doesnt result in additional work ?